File: Raon.Native.Tests/LexerTest.cpp
#include "pch.h"

#include "FrontEnd/Lexer.h"
#include "Utility/Error.h"

TEST(LexerTest, GetNextToken) {
    InitRaon();

    FLexer *lexer = nullptr;
    EXPECT_NO_THROW(lexer = CreateLexer(U16("1 + \"1\" * 2 - 3.0")));

    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_INTEGER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_PLUS);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_STRING);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_ASTERISK);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_INTEGER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_MINUS);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_REAL);

    EXPECT_NO_THROW(FreeLexer(lexer));

    lexer = CreateLexer(U16(R"(a = 1 + "1\" * 2 - 3.0
b = a / 10)"));

    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_IDENTIFIER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_ASSIGN);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_INTEGER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_PLUS);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_STRING);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_ASTERISK);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_INTEGER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_MINUS);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_REAL);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_EOL);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_IDENTIFIER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_ASSIGN);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_IDENTIFIER);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_SLASH);
    EXPECT_EQ(GetNextToken(lexer)->type, TOKEN_INTEGER);

    EXPECT_NO_THROW(FreeLexer(lexer));
}

TEST(LexerTest, Boolean) {
    FLexer *lexer = nullptr;
    FToken *token = nullptr;

    lexer = CreateLexer(U16("a = true"));
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_IDENTIFIER);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_ASSIGN);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_BOOL);
    EXPECT_EQ(token->value.boolean, true);
    EXPECT_NO_THROW(FreeLexer(lexer));

    lexer = CreateLexer(U16("a = false"));
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_IDENTIFIER);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_ASSIGN);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_BOOL);
    EXPECT_EQ(token->value.boolean, false);
    EXPECT_NO_THROW(FreeLexer(lexer));
}

TEST(LexerTest, Integer) {
    FLexer *lexer = nullptr;
    FToken *token = nullptr;

    lexer = CreateLexer(U16("a = 1"));
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_IDENTIFIER);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_ASSIGN);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_INTEGER);
    EXPECT_EQ(token->value.integer, 1);
    EXPECT_NO_THROW(FreeLexer(lexer));

    lexer = CreateLexer(U16("a = -2"));
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_IDENTIFIER);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_ASSIGN);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_MINUS);
    EXPECT_EQ((token = GetNextToken(lexer))->type, TOKEN_INTEGER);
    EXPECT_EQ(token->value.integer, 2);
    EXPECT_NO_THROW(FreeLexer(lexer));
}

File: Tests/Raon.Core.Tests/Syntastics/LexerTest.cs
using System;
using NUnit.Framework;
using Raon.Core.FrontEnd;

namespace Raon.Core.Tests.Syntastics;

[TestFixture]
public class LexerTest
{
    [Test]
    public void GetNextTokenTest()
    {
        using (Lexer lexer = new Lexer("1 + \"1\" * 2 - 3.0"))
        {
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_INTEGER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_PLUS));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_STRING));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_ASTERISK));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_INTEGER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_MINUS));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_REAL));
        }

        using (Lexer lexer = new Lexer(@"a = 1 + ""1\"" * 2 - 3.0
b = a / 10"))
        {
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_IDENTIFIER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_ASSIGN));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_INTEGER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_PLUS));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_STRING));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_ASTERISK));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_INTEGER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_MINUS));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_REAL));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_EOL));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_IDENTIFIER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_ASSIGN));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_IDENTIFIER));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_SLASH));
            Assert.That(lexer.GetNextToken().Type, Is.EqualTo(TokenType.TOKEN_INTEGER));
        }
    }
}